File: Android/pool/src/main/java/com/felipearpa/tyche/pool/joiner/PoolJoinerViewModel.kt
package com.felipearpa.tyche.pool.joiner

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.felipearpa.tyche.data.pool.application.GetPool
import com.felipearpa.tyche.data.pool.application.JoinPool
import com.felipearpa.tyche.data.pool.domain.JoinPoolInput
import com.felipearpa.ui.state.LoadState
import com.felipearpa.ui.state.SaveState
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class PoolJoinerViewModel(
    private val getPool: GetPool,
    private val joinPoolAction: JoinPool,
) : ViewModel() {
    private val _poolState =
        MutableStateFlow<LoadState<PoolModel>>(LoadState.Idle)
    val poolState = _poolState.asStateFlow()

    private val _joinPoolState =
        MutableStateFlow<SaveState<Unit>>(SaveState.Idle)
    val joinPoolState = _joinPoolState.asStateFlow()

    fun loadPool(poolId: String) {
        viewModelScope.launch {
            _poolState.emit(LoadState.Loading)
            val result = getPool.execute(poolId = poolId)
            result.onSuccess { pool ->
                _poolState.emit(LoadState.Loaded(pool.toPoolModel()))
            }.onFailure {
                _poolState.emit(LoadState.Failure(it))
            }
        }
    }

    fun joinPool(poolId: String, gamblerId: String) {
        viewModelScope.launch {
            _joinPoolState.emit(SaveState.Saving(Unit))
            val result =
                joinPoolAction.execute(JoinPoolInput(poolId = poolId, gamblerId = gamblerId))
            result.onSuccess {
                _joinPoolState.emit(SaveState.Saved(Unit))
            }.onFailure {
                _joinPoolState.emit(SaveState.Failure(Unit, it.asJoinPoolLocalizedException()))
            }
        }
    }
}

File: iOS/Pool/Sources/Pool/Joiner/PoolJoinerViewModel.swift
import Foundation
import DataPool
import UI
import ViewingState

public class PoolJoinerViewModel: ObservableObject {
    private let getPoolUseCase: GetPoolUseCase
    private let joinPoolUseCase: JoinPoolUseCase

    @Published private(set) var poolState: LoadState<PoolModel> = .idle
    @Published private(set) var joinPoolState: LoadState<Void> = .idle

    public init(getPoolUseCase: GetPoolUseCase, joinPoolUseCase: JoinPoolUseCase) {
        self.getPoolUseCase = getPoolUseCase
        self.joinPoolUseCase = joinPoolUseCase
    }

    @MainActor
    func loadPool(poolId: String) {
        Task {
            poolState = .loading
            let result = await getPoolUseCase.execute(poolId: poolId)

            switch result {
            case .success(let pool):
                poolState = .loaded(pool.toPoolModel())
            case .failure(let error):
                poolState = .failure(error)
            }
        }
    }

    @MainActor
    func joinPool(poolId: String, gamblerId: String) {
        Task {
            joinPoolState = .loading
            let result = await joinPoolUseCase.execute(
                joinPoolInput: JoinPoolInput(poolId: poolId, gamblerId: gamblerId)
            )

            switch result {
            case .success:
                joinPoolState = .loaded(())
            case .failure(let error):
                joinPoolState = .failure(error.asJoinPoolLocalizedError())
            }
        }
    }
}
